Sources/SourceKitLSP/SourceKitServer.swift

```diff
@@ -1702,73 +1662,112 @@ extension SourceKitServer {
+  /// Returns the result of a `DefinitionRequest` by running a `SymbolInfoRequest`, inspecting
+  /// its result and doing index lookups, if necessary.
+  ///
+  /// In contrast to `definition`, this does not fall back to sending a `DefinitionRequest` to the
+  /// toolchain language server.
+  private func indexBasedDefinition(
     _ req: DefinitionRequest,
     workspace: Workspace,
     languageService: ToolchainLanguageServer
-  ) async throws -> LocationsOrLocationLinksResponse? {
+  ) async throws -> [Location] {
     let symbols = try await languageService.symbolInfo(
       SymbolInfoRequest(
         textDocument: req.textDocument,
         position: req.position
       )
     )
-    let index = await self.workspaceForDocument(uri: req.textDocument.uri)?.index
+    guard let symbol = symbols.first else {
+      return []
+    }
+
+    if let symbol = symbols.first, let bestLocalDeclaration = symbol.bestLocalDeclaration, !(symbol.isDynamic ?? false)
+    {
+      // If we have a non-dynamic symbol, we don't need to do an index lookup
+      return [bestLocalDeclaration]
+    }
+
     // If this symbol is a module then generate a textual interface
-    if let symbol = symbols.first, symbol.kind == .module, let name = symbol.name {
-      return try await self.definitionInInterface(
+    if symbol.kind == .module, let name = symbol.name {
+      let interfaceLocation = try await self.definitionInInterface(
         req,
         moduleName: name,
         symbolUSR: nil,
         languageService: languageService
       )
+      return [interfaceLocation]
     }
 
-    let resolved = self.extractIndexedOccurrences(symbols: symbols, index: index, useLocalFallback: true) {
-      (usr, index) in
-      logger.info("performing indexed jump-to-def with usr \(usr)")
-      var occurs = index.occurrences(ofUSR: usr, roles: [.definition])
-      if occurs.isEmpty {
-        occurs = index.occurrences(ofUSR: usr, roles: [.declaration])
+    guard let index = await self.workspaceForDocument(uri: req.textDocument.uri)?.index else {
+      return []
+    }
+    guard let usr = symbol.usr else { return [] }
+    logger.info("performing indexed jump-to-def with usr \(usr)")
+    var occurances = index.occurrences(ofUSR: usr, roles: [.definition])
+    if occurances.isEmpty {
+      occurances = index.occurrences(ofUSR: usr, roles: [.declaration])
+    }
+    if symbol.isDynamic ?? false {
+      lazy var transitiveReceiverUsrs: [String] = transitiveSubtypeClosure(
+        ofUsrs: symbol.receiverUsrs ?? [],
+        index: index
+      )
+      occurances += occurances.flatMap {
+        let overrides = index.occurrences(relatedToUSR: $0.symbol.usr, roles: .overrideOf)
+        // Only contain overrides that are children of one of the receiver types or their subtypes.
+        return overrides.filter { override in
+          override.relations.contains(where: {
+            $0.roles.contains(.childOf) && transitiveReceiverUsrs.contains($0.symbol.usr)
```

Comment:
```suggestion
            $0.roles.contains(.childOf) && (transitiveReceiverUsrs.isEmpty || transitiveReceiverUsrs.contains($0.symbol.usr))
```

Reply:
Just noticed that as well. Changed it slightly so that `transitiveReceiverUsrs` is `nil` if not receivers were passed.